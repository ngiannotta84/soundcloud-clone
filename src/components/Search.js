import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { useParams } from "react-router-dom";
import Album from "./Album";
import Song from "./Song";
import Artist from "./Artist";
import getAlbums from "../requests/getAlbums";
import getSongs from "../requests/getSongs";
import getUsers from "../requests/getUsers";

const Search = ({ handleSetPlaylist }) => {
  const { name } = useParams();
  const [search, setSearch] = useState([]);

  useEffect(() => {
    (async () => {
      try {
        const albumResponse = await getAlbums({ name });
        const songResponse = await getSongs({ name });
        const artistResponse = await getUsers({ name });
        const data = [...albumResponse, ...songResponse, ...artistResponse];
        data.sort((a, b) => Date.parse(b.createdAt) - Date.parse(a.createdAt));
        setSearch(data);
      } catch (err) {
        setSearch([]);
      }
    })();
  }, [name]);

  return (
    <div className="feed">
      {search.length > 0 ? (
        search.map((data) => {
          if (data.AlbumId) {
            return (
              <Song
                artistName={data.Album.User.name}
                albumName={data.Album.name}
                albumArt={data.Album.url}
                songName={data.name}
                songAudio={data.url}
                handleSetPlaylist={handleSetPlaylist}
                key={`${data.id}${data.name}`}
              />
            );
          }
          if (data.UserId) {
            return (
              <Album
                artistName={data.User.name}
                albumName={data.name}
                albumArt={data.url}
                songs={data.Songs}
                handleSetPlaylist={handleSetPlaylist}
                key={`${data.id}${data.name}`}
                albumId={data.id}
              />
            );
          }
          if (data.email) {
            return (
              <Artist
                name={data.name}
                image={data.Albums[0]?.url}
                key={`${data.id}${data.name}`}
              />
            );
          }
          return null;
        })
      ) : (
        <h2>No results</h2>
      )}
    </div>
  );
};

Search.propTypes = {
  handleSetPlaylist: PropTypes.func.isRequired,
};

export default Search;
