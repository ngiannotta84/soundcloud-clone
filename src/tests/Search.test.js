import React from "react";
import { render, screen } from "@testing-library/react";
import Search from "../components/Search";

xtest("renders title", () => {
  render(<Search />);
  const linkElement = screen.getByText(/soundclone/i);
  expect(linkElement).toBeInTheDocument();
});
