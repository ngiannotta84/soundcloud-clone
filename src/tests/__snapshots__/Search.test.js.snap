// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search snapshots snapshot no user data 1`] = `
<DocumentFragment>
  <div
    class="feed"
  >
    <h2>
      No results
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Search snapshots snapshot with user data 1`] = `
<DocumentFragment>
  <div
    class="feed"
  >
    <div
      class="album"
    >
      <div
        class="album__info-container"
      >
        <img
          alt="album1 cover art"
          class="album__cover-art"
          src="art url 1"
        />
        <div
          class="album__info"
        >
          <h2
            class="album__artist-name"
          >
            <a
              href="/profile/artist name 1"
            >
              artist name 1
            </a>
          </h2>
          <h3
            class="album__album-name"
          >
            album1
          </h3>
        </div>
        <div
          class="album__buttons"
        >
          <button
            class="addSong-button addSong-large"
            type="button"
          >
            Add Album to queue
          </button>
          <button
            class="addSong-button addSong-large"
            type="button"
          >
            Play Album next
          </button>
        </div>
      </div>
      <li
        class="album__songs"
      >
        <ul
          class="album__song"
        >
          <h4
            class="album__song__song-name"
          >
            song1
          </h4>
          <div>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Add to queue
            </button>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Play next
            </button>
          </div>
        </ul>
        <ul
          class="album__song"
        >
          <h4
            class="album__song__song-name"
          >
            song2
          </h4>
          <div>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Add to queue
            </button>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Play next
            </button>
          </div>
        </ul>
      </li>
    </div>
    <div
      class="album"
    >
      <div
        class="album__info-container"
      >
        <img
          alt="album2 cover art"
          class="album__cover-art"
          src="art url 2"
        />
        <div
          class="album__info"
        >
          <h2
            class="album__artist-name"
          >
            <a
              href="/profile/artist name 2"
            >
              artist name 2
            </a>
          </h2>
          <h3
            class="album__album-name"
          >
            album2
          </h3>
        </div>
        <div
          class="album__buttons"
        >
          <button
            class="addSong-button addSong-large"
            type="button"
          >
            Add Album to queue
          </button>
          <button
            class="addSong-button addSong-large"
            type="button"
          >
            Play Album next
          </button>
        </div>
      </div>
      <li
        class="album__songs"
      >
        <ul
          class="album__song"
        >
          <h4
            class="album__song__song-name"
          >
            song3
          </h4>
          <div>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Add to queue
            </button>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Play next
            </button>
          </div>
        </ul>
        <ul
          class="album__song"
        >
          <h4
            class="album__song__song-name"
          >
            song4
          </h4>
          <div>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Add to queue
            </button>
            <button
              class="addSong-button album__song__button"
              type="button"
            >
              Play next
            </button>
          </div>
        </ul>
      </li>
    </div>
    <div
      class="song"
    >
      <div
        class="song__info-container"
      >
        <img
          alt="album 6tggj cover art"
          class="song__cover-art"
          src="url 0hgnid"
        />
        <div
          class="song__info"
        >
          <h2
            class="song__artist-name"
          >
            <a
              class="song__artist-name__link"
              href="/profile/name 7gbn"
            >
              name 7gbn
            </a>
          </h2>
          <h3
            class="song__song-name"
          >
            song 756453
          </h3>
        </div>
      </div>
      <div
        class="song__buttons"
      >
        <button
          class="addSong-button addSong-large"
          type="button"
        >
          Add to queue
        </button>
        <button
          class="addSong-button addSong-large"
          type="button"
        >
          Play next
        </button>
      </div>
    </div>
    <div
      class="song"
    >
      <div
        class="song__info-container"
      >
        <img
          alt="album jgjf6tggj cover art"
          class="song__cover-art"
          src="url 0hgnfgj6id"
        />
        <div
          class="song__info"
        >
          <h2
            class="song__artist-name"
          >
            <a
              class="song__artist-name__link"
              href="/profile/name 7gbhfgjfn"
            >
              name 7gbhfgjfn
            </a>
          </h2>
          <h3
            class="song__song-name"
          >
            song 7553
          </h3>
        </div>
      </div>
      <div
        class="song__buttons"
      >
        <button
          class="addSong-button addSong-large"
          type="button"
        >
          Add to queue
        </button>
        <button
          class="addSong-button addSong-large"
          type="button"
        >
          Play next
        </button>
      </div>
    </div>
    <div
      class="song"
    >
      <div
        class="song__info-container"
      >
        <img
          alt="albujfm 6tggj cover art"
          class="song__cover-art"
          src="url 0hgni754d"
        />
        <div
          class="song__info"
        >
          <h2
            class="song__artist-name"
          >
            <a
              class="song__artist-name__link"
              href="/profile/namefjrtu54 7gbn"
            >
              namefjrtu54 7gbn
            </a>
          </h2>
          <h3
            class="song__song-name"
          >
            song 756457553
          </h3>
        </div>
      </div>
      <div
        class="song__buttons"
      >
        <button
          class="addSong-button addSong-large"
          type="button"
        >
          Add to queue
        </button>
        <button
          class="addSong-button addSong-large"
          type="button"
        >
          Play next
        </button>
      </div>
    </div>
    <div
      class="artist"
    >
      <a
        class="artist__link"
        href="/profile/name 1"
      >
        <img
          alt="album by name 1"
          class="artist__cover-art"
          src="art url 1"
        />
        <h2
          class="artist__name"
        >
          name 1
        </h2>
      </a>
    </div>
    <div
      class="artist"
    >
      <a
        class="artist__link"
        href="/profile/name 2"
      >
        <img
          alt="album by name 2"
          class="artist__cover-art"
          src="art url 4"
        />
        <h2
          class="artist__name"
        >
          name 2
        </h2>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
